File: core/src/main/java/sk/magiksoft/wizard/Wizard.java

/***********************************************\
 *  Copyright (c) 2010 by Ing.Vladimir Hrusovsky *
 *  Sodalis 2007-2011                            *
 *  http://www.sodalis.sk                        *
 \***********************************************/
    
     
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package sk.magiksoft.wizard;

import javax.swing.*;
import java.awt.*;
import java.awt.event.ActionEvent;
import java.awt.event.KeyEvent;
import java.util.ResourceBundle;

/**
 * @author wladimiiir
 */
public class Wizard extends JDialog {
    private static final Icon PREVIOUS_ICON = new ImageIcon(Wizard.class.getResource("previous.png"));
    private static final Icon NEXT_ICON = new ImageIcon(Wizard.class.getResource("next.png"));
    private static final Icon FINISH_ICON = new ImageIcon(Wizard.class.getResource("finish.png"));
    private static final Icon CANCEL_ICON = new ImageIcon(Wizard.class.getResource("cancel.png"));

    public static final int CANCEL_ACTION = 0;
    public static final int FINISH_ACTION = 1;
    private JButton btnNextOrFinish = new JButton();
    private JButton btnPrevious = new JButton(new PreviousAction());
    private JButton btnCancel = new JButton(new CancelAction());
    private CardLayout cardLayout = new CardLayout();
    private JPanel pagePanel;
    private int currentPage = -1;
    private Component[] pages;
    private AbstractAction nextAction = new NextAction();
    private AbstractAction finishAction = new FinishAction();
    private int resultAction = CANCEL_ACTION;

    private static ResourceBundle strings = ResourceBundle.getBundle("sk.magiksoft.wizard.locale.wizard");

    public Wizard(Frame owner, Component[] pages) {
        super(owner, true);
        this.pages = pages;
        initComponents();
    }

    private void initComponents() {
        JPanel buttonPanel = new JPanel(new FlowLayout(FlowLayout.RIGHT));
        buttonPanel.add(btnPrevious);
        buttonPanel.add(btnNextOrFinish);
        buttonPanel.add(new JPanel());
        buttonPanel.add(btnCancel);
        pagePanel = new JPanel(cardLayout);

        for (int i = 0; i < pages.length; i++) {
            Component component = pages[i];
            pagePanel.add(component, String.valueOf(i));
        }
        cardLayout.show(pagePanel, String.valueOf(0));
        currentPage = 0;

        this.setLayout(new BorderLayout());
        this.getContentPane().add(pagePanel, BorderLayout.CENTER);
        this.getContentPane().add(buttonPanel, BorderLayout.SOUTH);
        this.setSize(300, 200);

        setupButtons();

        ((JComponent) getContentPane()).getInputMap(JComponent.WHEN_IN_FOCUSED_WINDOW).put(KeyStroke.getKeyStroke(KeyEvent.VK_ESCAPE, 0), "ESCAPE");
        ((JComponent) getContentPane()).getActionMap().put("ESCAPE", new CancelAction());
    }

    public int showWizard() {
        currentPage = 0;
        cardLayout.show(pagePanel, String.valueOf(currentPage));
        setupButtons();
        this.setVisible(true);
        return resultAction;
    }

    private void setupButtons() {
        if (currentPage == 0) {
            btnPrevious.setEnabled(false);
        } else {
            btnPrevious.setEnabled(true);
        }

        if (currentPage == (pages.length - 1)) {
            btnNextOrFinish.setAction(finishAction);
        } else {
            btnNextOrFinish.setAction(nextAction);
        }
    }

    private class NextAction extends AbstractAction {

        public NextAction() {
            super(strings.getString("next"), NEXT_ICON);
        }

        @Override
        public void actionPerformed(ActionEvent e) {
            cardLayout.show(pagePanel, String.valueOf(++currentPage));
            setupButtons();
        }
    }

    private class PreviousAction extends AbstractAction {

        public PreviousAction() {
            super(strings.getString("previous"), PREVIOUS_ICON);
        }

        @Override
        public void actionPerformed(ActionEvent e) {
            cardLayout.show(pagePanel, String.valueOf(--currentPage));
            setupButtons();
        }
    }

    private class FinishAction extends AbstractAction {

        public FinishAction() {
            super(strings.getString("finish"), FINISH_ICON);
        }

        @Override
        public void actionPerformed(ActionEvent e) {
            resultAction = FINISH_ACTION;
            setVisible(false);
        }
    }

    private class CancelAction extends AbstractAction {

        public CancelAction() {
            super(strings.getString("cancel"), CANCEL_ICON);
        }

        @Override
        public void actionPerformed(ActionEvent e) {
            resultAction = CANCEL_ACTION;
            setVisible(false);
        }
    }
}

File: core/src/main/java/sk/magiksoft/sodalis/core/ui/wizard/Wizard.scala
package sk.magiksoft.sodalis.core.ui.wizard

import java.awt.{BorderLayout, Frame}
import sk.magiksoft.sodalis.core.locale.LocaleManager
import sk.magiksoft.sodalis.core.factory.IconFactory
import javax.swing.{JPanel, JDialog}
import sk.magiksoft.sodalis.core.utils.UIUtils
import swing._
import event.Event

/**
 * @author wladimiiir
 * @since 2011/2/18
 */
class Wizard(owner: Frame, title: String, page: Page) extends JDialog(owner, title, true) with Publisher {
  private var currentPage = page
  private val previousButton = new Button(previousAction)
  private val nextButton = new Button
  private val centerPanel = new JPanel(new BorderLayout)

  initComponents
  UIUtils.makeISDialog(this)

  private def initComponents = {
    val buttonPanel = new FlowPanel(FlowPanel.Alignment.Right)(
      previousButton,
      nextButton,
      new Separator,
      new Button(new Action(LocaleManager.getString("cancelAction")) {
        icon = IconFactory.getInstance.getIcon("cancel")

        def apply() = {
          setVisible(false)
          publish(new WizardCancelled(Wizard.this))
        }
      })
    )

    setLayout(new BorderLayout)
    add(centerPanel, BorderLayout.CENTER)
    add(buttonPanel.peer, BorderLayout.SOUTH)
  }

  private def refresh = {
    centerPanel.removeAll
    centerPanel.add(currentPage.getComponent.peer, BorderLayout.CENTER)
    centerPanel.revalidate
    centerPanel.repaint()
    previousButton.enabled = currentPage.getPreviousPage.isDefined
    nextButton.action = currentPage.getNextPage match {
      case Some(_) => nextAction
      case None => finishAction
    }
  }

  def showWizard {
    currentPage = page
    refresh
    setVisible(true)
  }

  object previousAction extends Action(LocaleManager.getString("previous")) {
    icon = IconFactory.getInstance.getIcon("previous")

    def apply = {
      currentPage = currentPage.getPreviousPage.get
      refresh
    }
  }

  object nextAction extends Action(LocaleManager.getString("next")) {
    icon = IconFactory.getInstance.getIcon("next")

    def apply = {
      currentPage = currentPage.getNextPage.get
      refresh
    }
  }

  object finishAction extends Action(LocaleManager.getString("finish")) {
    icon = IconFactory.getInstance.getIcon("finish")

    def apply = {
      setVisible(false)
      publish(new WizardFinished(Wizard.this))
    }
  }

}

case class WizardFinished(wizard: Wizard) extends Event

case class WizardCancelled(wizard: Wizard) extends Event
